File: android/app/src/main/java/com/patron/snaglite/download/resolvers/ResolverPipeline.kt
package com.patron.snaglite.download.resolvers

import android.net.Uri
import okhttp3.OkHttpClient

class ResolverPipeline {

    private val http: OkHttpClient = HttpFetch.newClient()

    private val resolvers: List<IResolver> = listOf(
        ResolverA(),
        ResolverB(),
        ResolverC(),
    )

    fun match(url: String): IResolver? {
        val host = Uri.parse(url).host?.lowercase() ?: return null
        return resolvers.firstOrNull { it.matches(host) }
    }

    suspend fun resolve(url: String): ResolvedMedia? {
        val r = match(url) ?: return null
        return r.resolve(url, http)
    }

    suspend fun unwrapIfNeeded(url: String): String {
        val host = Uri.parse(url).host?.lowercase() ?: return url
        if (!IframeUnwrapper.shouldUnwrap(host)) return url
        return IframeUnwrapper.unwrap(url, http) ?: url
    }
}

File: src/SnagLite/Services/Resolvers/ResolverPipeline.cs
namespace SnagLite.Services.Resolvers;

public sealed class ResolverPipeline
{
    private readonly IReadOnlyList<IResolver> _resolvers;
    private readonly HttpClient _http;

    public ResolverPipeline()
    {
        var handler = new HttpClientHandler
        {
            AllowAutoRedirect = true,
            AutomaticDecompression =
                System.Net.DecompressionMethods.GZip |
                System.Net.DecompressionMethods.Deflate |
                System.Net.DecompressionMethods.Brotli,
        };
        _http = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(30) };

        _resolvers = new IResolver[]
        {
            new ResolverA(),
            new ResolverB(),
            new ResolverC(),
        };
    }

    public IResolver? Match(Uri url) => _resolvers.FirstOrDefault(r => r.Matches(url));

    public async Task<ResolvedMedia?> ResolveAsync(Uri url, CancellationToken ct)
    {
        var r = Match(url);
        if (r is null) return null;
        return await r.ResolveAsync(url, _http, ct);
    }
}
